### PR TITLE
Andrew

### PR DESCRIPTION
**Pull request by Andrew-JG-Ha**
Major Changes

### CODE REVIEW

File: pong/gameServer.h [hunk 1]
```diff
@@ -2,24 +2,38 @@
...
+    MatrixHardware matrix;
+} OutputHardware;
 
-typedef struct gameServer {
+typedef struct GameServer {
     pthread_t pId;
     int board[BOARD_SIZE][BOARD_SIZE];
     int scoreLeft;
     int scoreRight;
     int directionX;
```
Comment (NickMoo5) [inline pong/gameServer.h]: can we rename these direction vars to smth that indicates that it's the ball direction

File: pong/hardware/lcd.c [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    char configPinCmd[MAX_LEN];
+    printf("fileDIr is: %s\n\n", fileDir);
+
+    sprintf(configPinCmd, "config-pin %s gpio", pin); 
+    runCommand(configPinCmd);
+    if (readFile(fileDir, "direction", buffer) == -1) {
+        writeToFile(GPIO_FILE_DIR, "export", pinNumber);
+    }
+    writeToFile(fileDir, "direction", "out");
+    //writeToFile(fileDir, "active_low", "0");
```
Comment (NickMoo5) [inline pong/hardware/lcd.c]: doesn't hurt to set it just in case

**Review by NickMoo5 — COMMENTED** [review]
MOO

File: pong/gameServer.c [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    newGame->player1 = player1;
+    newGame->player2 = player2; 
+    newGame->lcdScreen = generateLcd(hw.ledScreen);
+    newGame->matrixHardware = hw.matrix;
+}
+
+void destroyGameServer(GameServer* gameServer) {
+    destroyLcd(gameServer->lcdScreen);
+    gameServer->lcdScreen = NULL;
+    free(gameServer);
```
Comment (ShakeebZ) [inline pong/gameServer.c:31]: do gameServer = NULL; after this line

File: pong/hardware/lcd.c [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    sprintf(newLcdScreen->rsPinFilePath, "%s%s/", GPIO_FILE_DIR, lcdHardware.rsPin.gpioPin);
+    initPin(newLcdScreen->rsPinFilePath, lcdHardware.rsPin.pin, lcdHardware.rsPin.pinNumber);
+    
+    /**
+     * The following segment of code is adapted from 
+     * https://opencoursehub.cs.sfu.ca/mba200/grav-cms/ensc351/links/files/2022-student-howtos/16x2CharacterLCDThroughGPIO.pdf
+    */
+
+    // Special Function Set 1.
+	write4BitsToLcd(newLcdScreen, 0x03); // 0011
```
Comment (ShakeebZ) [inline pong/hardware/lcd.c:39]: QOL: The write4BitsToLcd function calls are indented by 1 more tab than they should be
fix: shift tab the relevant lines

File: pong/gameServer.h [hunk 5]
```diff
@@ -2,24 +2,38 @@
...
+
 #define BOARD_SIZE 16
 #define PADDLE_SIZE 3
 
 /**
  * Game server would read the playerClient desired move ie up/down/nothing and update game state
  * and update the display
 */
+typedef struct OutputHardware {
+    LcdHardware ledScreen;
```
Comment (ShakeebZ) [inline pong/gameServer.h]: should be lcdScreen not ledScreen

File: pong/pong.c [hunk 6]
```diff
@@ -0,0 +1,148 @@
+#include "playerClient.h"
+#include "gameServer.h"
+
+#define USER_BUTTON "/sys/class/gpio/gpio72/"
+#define MAX_LENGTH 1024
+
+#define matrixAddress1
```
Comment (ShakeebZ) [inline pong/pong.c]: this isn't initialized to something, looks like a ghost that should be removed

**Review by ShakeebZ — COMMENTED** [review]
Review comments left in review

**Review by danielqyu — APPROVED** [review]

**Review by ShakeebZ — APPROVED** [review]
Changes implemented, am happy

**Pull request merged**